contracts/user/SharedConsumer.sol

```diff
@@ -0,0 +1,314 @@
+        } else if (playType == PlayType.Roll) {
+            (uint32 bunch, ) = abi.decode(params, (uint32, uint32));
+            gasLimit = ROLL_CALLBACK_GAS_BASE + bunch * 100;
+        }
+    }
+
+    struct FeeConfig {
+        uint16 flatFeePromotionGlobalPercentage;
+        bool isFlatFeePromotionEnabledPermanently;
+        uint256 flatFeePromotionStartTimestamp;
+        uint256 flatFeePromotionEndTimestamp;
+    }
+
+    function _getFlatFeeConfig() internal view returns (FeeConfig memory feeConfig) {
+        {
+            (, bytes memory point) =
+            // solhint-disable-next-line avoid-low-level-calls
+             address(adapter).staticcall(abi.encodeWithSelector(IAdapter.getFlatFeeConfig.selector));
+            uint16 flatFeePromotionGlobalPercentage;
+            bool isFlatFeePromotionEnabledPermanently;
+            uint256 flatFeePromotionStartTimestamp;
+            uint256 flatFeePromotionEndTimestamp;
+            // solhint-disable-next-line no-inline-assembly
+            assembly {
+                flatFeePromotionGlobalPercentage := mload(add(point, 320))
+                isFlatFeePromotionEnabledPermanently := mload(add(point, 352))
+                flatFeePromotionStartTimestamp := mload(add(point, 384))
+                flatFeePromotionEndTimestamp := mload(add(point, 416))
+            }
+            feeConfig = FeeConfig(
+                flatFeePromotionGlobalPercentage,
+                isFlatFeePromotionEnabledPermanently,
+                flatFeePromotionStartTimestamp,
+                flatFeePromotionEndTimestamp
+            );
+        }
+    }
+
+    function rollDice(uint32 bunch, uint32 size, uint64 subId, uint256 seed, uint16 requestConfirmations)
+        external
+        payable
+        returns (bytes32 requestId)
+    {
+        bytes memory calcParams = abi.encode(bunch, size);
+        uint32 gasLimit = _calculateGasLimit(PlayType.Roll, calcParams);
+        if (gasLimit > MAX_GAS_LIMIT) {
+            revert GasLimitTooBig(gasLimit, MAX_GAS_LIMIT);
+        }
+        if (requestConfirmations == 0) {
+            (requestConfirmations,,,,,,) = IAdapter(adapter).getAdapterConfig();
+        }
+        subId = _fundSubId(PlayType.Roll, subId, calcParams);
+        bytes memory params;
+        params = abi.encode(bunch);
+        requestId = _rawRequestRandomness(
+            RequestType.RandomWords, params, subId, seed, requestConfirmations, gasLimit, tx.gasprice * 3
+        );
+        emit RollDiceRequest(msg.sender, subId, requestId, bunch, size, msg.value, seed, requestConfirmations);
+        pendingRequests[requestId] = RequestData(PlayType.Roll, calcParams);
+    }
+
+    function drawTickets(
+        uint32 totalNumber,
+        uint32 winnerNumber,
+        uint64 subId,
+        uint256 seed,
+        uint16 requestConfirmations
+    ) external payable returns (bytes32 requestId) {
+        bytes memory calcParams = abi.encode(totalNumber, winnerNumber);
```

Comment:
Insert validity check on totalNumber and winnerNumber.

Reply:
The point is, if the request succeeds, then fulfillment succeeds. We shouldn't let things break on _fulfillRandomness or _fulfillRandomWords.

Reply:
calcParams -> requestParams